src/wallet/mod.rs

```diff
@@ -243,6 +244,19 @@ where
         })
     }
 
+    /// This checks the checksum within [`BatchDatabase`] twice (if needed). The first time with the
+    /// actual checksum, and the second time with the checksum of `descriptor+checksum`. The second
+    /// check is necessary for backwards compatibility of a checksum-inception bug.
+    fn db_checksum(db: &mut D, desc: &str, kind: KeychainKind) -> Result<(), Error> {
+        let checksum = get_checksum_bytes(desc, true)?;
+        if db.check_descriptor_checksum(kind, checksum).is_ok() {
+            return Ok(());
+        }
+
+        let checksum_inception = get_checksum_bytes(desc, false)?;
+        db.check_descriptor_checksum(kind, checksum_inception)
+    }
```

Comment:
Instead of creating another function here, is it possible to move this logic inside `check_descriptor_checksum` itself? This hacky bug fix doesn't need to be so up high in the library. And can be confusing later on.

Reply:
`check_descriptor_checksum` is part of a trait.

Reply:
Ah right.. Trait.. 😄